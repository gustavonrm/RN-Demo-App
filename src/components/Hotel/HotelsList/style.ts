import { StyleSheet } from 'react-native';

const style = StyleSheet.create({
  container: {
    marginHorizontal: 20,
  },
});

export default style;
