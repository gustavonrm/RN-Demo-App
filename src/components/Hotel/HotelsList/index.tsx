import React, { useMemo } from 'react';
import { FlatList } from 'react-native';
import { Hotel } from '../types/hotel';
import HotelCard from '../HotelCard';
import withLoader from '../../HOC/withLoader';
import { useGetHotelsQuery } from '../../../redux/apis/hotels.api';
import style from './style';
import withError from '../../HOC/withError';
import { selectHotels } from '../../../redux/slices/hotels.slice';
import { useSelector } from 'react-redux';
import { selectFilters } from '../../../redux/slices/filters.slice';
import ErrorView from '../../common/ErrorView';

type Filters = {
  stars?: number;
  price?: string;
};

const HotelsList = () => {
  const hotels = useSelector(selectHotels);
  const filters = useSelector(selectFilters) as Filters;
  // const [filtering, setFiltering] = useState(false); // Not added because the filtering is very fast and the modal also hides this
  const filteredHotels = useMemo(() => {
    const filterKeys = ['stars', 'price'];
    const hasFilters = filterKeys.some((key) => key in filters);

    let filtered = hotels.filter((hotel) => {
      if (!hasFilters) {
        return true;
      }

      const matchesStars = filters.stars !== undefined ? hotel.stars === filters.stars : true;

      return matchesStars;
    });

    if (filters.price) {
      filtered = filtered.sort((a: Hotel, b: Hotel) => {
        if (filters.price === 'desc') {
          return b.price - a.price;
        } else if (filters.price === 'asc') {
          return a.price - b.price;
        }
        return 0;
      });
    }

    return filtered;
  }, [hotels, filters]);

  const renderItem = ({ item }: { item: Hotel }) => {
    return <HotelCard hotel={item} />;
  };

  return (
    <FlatList
      contentContainerStyle={{ flexGrow: 1 }}
      style={style.container}
      showsVerticalScrollIndicator={false}
      data={filteredHotels}
      renderItem={renderItem}
      keyExtractor={(item) => item.id}
      ListEmptyComponent={<ErrorView message={'No hotels found'} />}
    />
  );
};

export default withLoader(withError(HotelsList, selectHotels), useGetHotelsQuery);
