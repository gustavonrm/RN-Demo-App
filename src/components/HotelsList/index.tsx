import React from 'react';
import { FlatList } from 'react-native';
import { Hotel } from '../types/hotel';
import HotelCard from '../HotelCard';
import withLoader from '../HOC/withLoader';
import { useGetHotelsQuery } from '../../redux/apis/hotels.api';
import style from './style';

type HotelListProps = {
  data: Hotel[];
};

const HotelsList = ({ data }: HotelListProps) => {
  const renderItem = ({ item }: { item: Hotel }) => <HotelCard hotel={item} />;

  return (
    <FlatList
      style={style.container}
      showsVerticalScrollIndicator={false}
      data={data}
      renderItem={renderItem}
      keyExtractor={(item) => item.id}
    />
  );
};

export default withLoader(HotelsList, useGetHotelsQuery);
