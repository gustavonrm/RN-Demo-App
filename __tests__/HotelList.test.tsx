import React from 'react';
import 'react-native';
import { cleanup, screen, waitForElementToBeRemoved } from '@testing-library/react-native';
import HotelsList from '../src/components/HotelsList';
import { renderWithProviders } from './utils/test-utils';
import { http, HttpResponse, delay } from 'msw';
import { setupServer } from 'msw/native';
import { hotels } from './mocks/hotels';

export const handlers = [
  http.get('https://run.mocky.io/v3/b4593f68-9e29-47a8-bb6d-8137a665ee15', async () => {
    await delay(150);
    return HttpResponse.json(hotels);
  }),
];
export const server = setupServer(...handlers);

// Enable API mocking before tests.
beforeAll(() =>{
  server.listen();
  jest.useFakeTimers();
});

// Reset any runtime request handlers we may add during the tests.
afterEach(() =>{
  server.resetHandlers();
  jest.restoreAllMocks();
  cleanup();
});

// Disable API mocking after the tests are done.
afterAll(() => server.close());

describe('HotelList Element', () => {
  it('renders hotel list correctly', async () => {
    renderWithProviders(<HotelsList />);
    expect(screen.getByTestId('loaderTestId')).toBeTruthy();
    await waitForElementToBeRemoved(() => screen.getByTestId('loaderTestId'));

    expect(await screen.getAllByTestId('hotelCardTestId').length).toBe(10);
  });
});
