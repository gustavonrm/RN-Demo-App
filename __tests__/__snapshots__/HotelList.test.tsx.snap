// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HotelList Element error snapshot test 1`] = `
<ActivityIndicator
  color="#E12C70"
  size="large"
  style={
    {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
  testID="loaderTestId"
/>
`;
